besskge/pipeline.py

```diff
@@ -0,0 +1,279 @@
+                )
+                # Global indices of entities scored in the step
+                batch_idx.append(
+                    self.bess_module.sharding.shard_and_idx_to_entity[
+                        :, ent_slice
+                    ].flatten()
+                )
+                inp = {k: v.flatten(end_dim=1) for k, v in batch.items()}
+                inp.update(dict(step=step))
+                batch_res.append(self.poptorch_module(**inp))
+            batch_scores = torch.concat(batch_res, dim=-1)
+            # Filter out padding scores
+            batch_scores_filt = batch_scores[triple_mask.flatten()][
+                :, np.unique(np.concatenate(batch_idx), return_index=True)[1]
+            ][:, : self.bess_module.sharding.n_entity]
+            # Scores of positive triples
+            true_scores = batch_scores_filt[
+                torch.arange(batch_scores_filt.shape[0]),
+                ground_truth[triple_mask],
+            ]
+            if self.filter_triples is not None:
+                # Filter for triples in batch
+                batch_filter = get_entity_filter(
+                    self.triples[triple_id[triple_mask]],
+                    self.filter_triples,
+                    filter_mode=self.corruption_scheme,
+                )
+                batch_scores_filt[batch_filter[:, 0], batch_filter[:, 1]] = -torch.inf
+
+            if self.evaluation:
+                assert (
+                    ground_truth is not None
+                ), "Evaluation requires providing ground truth entities"
+                # If not already masked, mask scores of true triples
+                # to compute metrics
+                batch_scores_filt[
+                    torch.arange(batch_scores_filt.shape[0]),
+                    ground_truth[triple_mask],
+                ] = -torch.inf
+                batch_ranks = self.evaluation.ranks_from_scores(
+                    true_scores, batch_scores_filt
+                )
+                metrics.append(self.evaluation.dict_metrics_from_ranks(batch_ranks))
+                if self.evaluation.return_ranks:
+                    ranks.append(batch_ranks)
+            if self.return_scores:
+                # Restore positive scores in the returned scores
+                batch_scores_filt[
+                    torch.arange(batch_scores_filt.shape[0]),
+                    ground_truth[triple_mask],
+                ] = true_scores
+                scores.append(batch_scores_filt)
+
+        out = dict()
+        if scores:
+            out["scores"] = torch.concat(scores, dim=0)
+        if ids:
+            out["triple_idx"] = torch.concat(ids, dim=0)
+        if self.evaluation:
+            concat_metrics = dict()
+            for m in metrics[0].keys():
+                concat_metrics[m] = torch.concat(
+                    [met[m].reshape(-1) for met in metrics]
+                )
+            out["metrics"] = concat_metrics  # type: ignore
```

Comment:
not sure these are really relevant? It's sums of metrics per batch I believe?

Reply:
I see your point. What is returned here is based on the `reduction` of the `Evaluation` class that is passed to the pipeline. If `reduction="sum"`, I agree that returning the sums per batch is not very useful in general (maybe we should just sum all together?), but there is also the option `reduction="none"`, which would return metrics for each triple - and that I think could be useful to support?

Reply:
Ah yes, you're right of course! With unreduced results that certainly can be useful

Reply:
I added in [247d88b](https://github.com/graphcore-research/bess-kge/pull/33/commits/247d88b4c3a38d816eb66f409e6887d111671d5f) a final reduction of metrics over all batches (so, if `reduction=sum`, in the output we get the sum over all triples, which seems more coherent).